Initial Snowflake integration

cyclux opened the pull request:
This pull request introduces a new Snowflake integration module for the Jaffle Shop dataset. It provides a robust, idempotent workflow for bootstrapping Snowflake infrastructure, ingesting CSV data from S3, and preparing weekly sales forecasting data for use with getML. The implementation features modular Python scripts, externalized SQL queries, and comprehensive logging and error handling. The workflow is automated via a new GitHub Actions CI pipeline.

**Infrastructure and Workflow Automation**
* Added a new GitHub Actions workflow (`.github/workflows/snowflake-test.yml`) to automate Python linting, formatting, type checking, and testing for the Snowflake integration, including coverage reporting and support for multiple Python versions.

**Snowflake Infrastructure Bootstrapping**
* Implemented `bootstrap.py` to create Snowflake warehouses and databases if they do not exist, using idempotent SQL and externalized queries (`create_warehouse.sql`, `create_database.sql`). [[1]](diffhunk://#diff-b37900d6828e9c7284db83b3cc4aaa73cf5ba2fca2186b44e06cdca57d75017dR1-R69) [[2]](diffhunk://#diff-a25d957c5c0b286fe59c3de41567bf0c4836a32e517a2ab9c07d5a61a2673a20R1-R7) [[3]](diffhunk://#diff-0080832b6b5c409bed325009f227af00753b612b99afd40c9d12495796e96943R1-R2)

**Data Ingestion Pipeline**
* Developed `data/ingestion.py` to ingest Jaffle Shop CSV data from S3 into Snowflake's RAW schema using external stages and native COPY INTO commands, with transaction management and error handling. [[1]](diffhunk://#diff-447c508a40df55639e012b970df249b558eee2b04591818f33474ad87dfca115R1-R148) [[2]](diffhunk://#diff-e72f0be46a412e2deeeeaa1934408144cfd5b6961c08b988e752b0834eca96dcR1-R9) [[3]](diffhunk://#diff-d36325015402cc45e602f45570208dd9ad9449eedd747a889b0507a31cc844c7R1-R2)

**Data Preparation for Feature Store**
* Added `data/preparation.py` to create weekly sales population tables per store, calculate forecasting targets, perform schema validation, and run data quality checks, leveraging externalized SQL for maintainability. [[1]](diffhunk://#diff-8d0d3a21ecfdb4d57ab79af9bf5f3dc3903203d73e3b80da74946e228e1c7491R1-R293) [[2]](diffhunk://#diff-a733db1ccc3d3edd3ddb3d17a84c2b8cf1474a04beae6575c4b0a9c0e5fb98a7R1-R13)

**Modular Utilities and API**
* Introduced `data/__init__.py` and `_sql_loader.py` to provide a clean API and internal SQL file loading utilities for the integration package. [[1]](diffhunk://#diff-6687cf2f2f14d720aeb2a0d59c29bbe02b1adc952e5f2b543f5f99927a3be95dR1-R33) [[2]](diffhunk://#diff-302453646013021af6743f4a71bc0bf14b8ff78a8f56568e1c513e545c58af73R1-R34)

cyclux committed:
Initial Snowflake integration

srnnkls commented on integration/snowflake/data/sql/preparation/create_population_weekly_by_store.sql:
```diff
@@ -0,0 +1,56 @@
+-- Create weekly snapshots per store
```
```sql
CREATE TABLE PREPARED.weekly_stores AS
WITH store_activity AS (
    SELECT 
        s.id as store_id,
        s.name as store_name,
        s.opened_at,
        DATE_TRUNC('week', s.opened_at) + INTERVAL '7 days' as first_full_week,
        MIN(o.ordered_at) as first_order_date,
        MAX(o.ordered_at) as last_order_date,
        DATE_TRUNC('week', MIN(o.ordered_at)) as first_order_week,
        DATE_TRUNC('week', MAX(o.ordered_at)) as last_order_week
    FROM RAW.raw_stores s
    LEFT JOIN RAW.raw_orders o ON o.store_id = s.id
    GROUP BY s.id, s.name, s.opened_at
),

all_weeks AS (
    SELECT DISTINCT 
        DATE_TRUNC('week', ordered_at) as reference_date
    FROM RAW.raw_orders
    WHERE ordered_at IS NOT NULL
),

store_weeks AS (
    SELECT 
        sa.store_id,
        sa.store_name,
        w.reference_date,
        sa.opened_at,
        sa.first_full_week,
        sa.first_order_week,
        sa.last_order_week
    FROM store_activity sa
    CROSS JOIN all_weeks w
    WHERE w.reference_date >= sa.opened_at
      AND w.reference_date < sa.last_order_week
)

SELECT 
    store_id,
    store_name,
    reference_date,
    EXTRACT(year FROM reference_date) as year,
    EXTRACT(month FROM reference_date) as month,
    EXTRACT(week FROM reference_date) as week_number,
    DATEDIFF('day', opened_at, reference_date) as days_since_open,
    reference_date >= first_full_week as is_full_week_after_opening,
    first_order_week IS NOT NULL 
        AND reference_date >= first_order_week
        AND reference_date < last_order_week as has_order_activity,
    DATEDIFF('day', opened_at, reference_date) >= 7 as has_min_history
FROM store_weeks
ORDER BY reference_date, store_id;
```

cyclux committed:
Major refactor

cyclux committed:
Refactor code structure for improved readability and maintainability

cyclux committed:
feat(snowflake): Implement Snowflake infrastructure bootstrap and session management

- Added bootstrap functionality to ensure the existence of Snowflake warehouse and database.
- Created SnowflakeSettings class for managing authentication and connection settings.
- Implemented session management for Snowflake using Snowpark.
- Developed SQL loading utilities for dynamic SQL execution.
- Added SQL scripts for creating schemas, tables, and stages, as well as data ingestion processes.
- Prepared a script to load and prepare Jaffle Shop data for integration with getML Feature Store.
